### ISSUE TITLE
source_spec.rb: "with color" tests only work on CI that return TTY=true (e.g. Travis, not AppVeyor or Azure Pipelines)

### ISSUE THREAD

**Issue by janpio** [bundler/bundler#6900]
When creating #6899 I noticed some 3 tests failing for both macOS and Linux on both AppVeyor and Azure Pipelines that should not really fail:

```
2019-01-09T21:37:36.5421043Z Failures:
2019-01-09T21:37:36.5421043Z 
2019-01-09T21:37:36.5421243Z   1) Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with a different version with color should return a string with the spec name and version and locked spec version
2019-01-09T21:37:36.5421494Z      Failure/Error: expect(subject.version_message(spec)).to eq("nokogiri >= 1.6\e[32m (was < 1.5)\e[0m")
2019-01-09T21:37:36.5421644Z 
2019-01-09T21:37:36.5422113Z        expected: "nokogiri >= 1.6\e[32m (was < 1.5)\e[0m"
2019-01-09T21:37:36.5422288Z             got: "nokogiri >= 1.6 (was < 1.5)"
2019-01-09T21:37:36.5422423Z 
2019-01-09T21:37:36.5422594Z        (compared using ==)
2019-01-09T21:37:36.5423477Z      # ./spec/bundler/source_spec.rb:63:in `block (7 levels) in <top (required)>'
2019-01-09T21:37:36.5423891Z      # ./spec/spec_helper.rb:122:in `block (2 levels) in <top (required)>'
2019-01-09T21:37:36.5424038Z 
2019-01-09T21:37:36.5424214Z   2) Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with a more recent version with color should return a string with the locked spec version in yellow
2019-01-09T21:37:36.5424407Z      Failure/Error: expect(subject.version_message(spec)).to eq("nokogiri 1.6.1\e[33m (was 1.7.0)\e[0m")
2019-01-09T21:37:36.5424523Z 
2019-01-09T21:37:36.5424667Z        expected: "nokogiri 1.6.1\e[33m (was 1.7.0)\e[0m"
2019-01-09T21:37:36.5424799Z             got: "nokogiri 1.6.1 (was 1.7.0)"
2019-01-09T21:37:36.5424935Z 
2019-01-09T21:37:36.5425056Z        (compared using ==)
2019-01-09T21:37:36.5425970Z      # ./spec/bundler/source_spec.rb:82:in `block (7 levels) in <top (required)>'
2019-01-09T21:37:36.5426708Z      # ./spec/spec_helper.rb:122:in `block (2 levels) in <top (required)>'
2019-01-09T21:37:36.5427033Z 
2019-01-09T21:37:36.5427216Z   3) Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with an older version with color should return a string with the locked spec version in green
2019-01-09T21:37:36.5427399Z      Failure/Error: expect(subject.version_message(spec)).to eq("nokogiri 1.7.1\e[32m (was 1.7.0)\e[0m")
2019-01-09T21:37:36.5427540Z 
2019-01-09T21:37:36.5427674Z        expected: "nokogiri 1.7.1\e[32m (was 1.7.0)\e[0m"
2019-01-09T21:37:36.5427811Z             got: "nokogiri 1.7.1 (was 1.7.0)"
2019-01-09T21:37:36.5427944Z 
2019-01-09T21:37:36.5428075Z        (compared using ==)
2019-01-09T21:37:36.5428437Z      # ./spec/bundler/source_spec.rb:95:in `block (7 levels) in <top (required)>'
2019-01-09T21:37:36.5428820Z      # ./spec/spec_helper.rb:122:in `block (2 levels) in <top (required)>'
2019-01-09T21:37:36.5526797Z 
2019-01-09T21:37:36.5526986Z Finished in 20 minutes 19 seconds (files took 5.19 seconds to load)
2019-01-09T21:37:36.5527042Z 2509 examples, 3 failures, 1 pending
2019-01-09T21:37:36.5527185Z 
2019-01-09T21:37:36.5527278Z Failed examples:
2019-01-09T21:37:36.5527323Z 
2019-01-09T21:37:36.5527432Z rspec ./spec/bundler/source_spec.rb:62 # Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with a different version with color should return a string with the spec name and version and locked spec version
2019-01-09T21:37:36.5527644Z rspec ./spec/bundler/source_spec.rb:81 # Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with a more recent version with color should return a string with the locked spec version in yellow
2019-01-09T21:37:36.5527798Z rspec ./spec/bundler/source_spec.rb:94 # Bundler::Source#version_message when there are locked gems that contain the relevant gem spec with an older version with color should return a string with the locked spec version in green
```

From the test failure output, we know that the command should have returned a "color code" in the message, but didn't.

Looking at the "shell creation code", gave some pointers:
https://github.com/janpio/bundler/blob/4c81ec16df2a7623ec93fd8f6169bd7e495781a7/lib/bundler/ui/shell.rb#L12-L19
`!$stdout.tty?` is used to decide if to fall back to a basic, uncolored Shell.

This can be simulated locally by piping the test run to e.g. `cat`: `ruby bin/rspec spec/bundler/source_spec.rb:63 | cat`

By running a simple `tty.rb` with content `puts $stdout.isatty` I confirmed that both AppVeyor and Azure Pipelines return `false`, but Travis returns `true`.

So the fix is clear: We have to make the tests return `tty? = true` for these tests to make sure the Shell will return colored output.

I will submit a PR to make these tests more stable.

**Comment by janpio** [comment]
Well, `allow($stdout).to receive(:tty?).and_return(true)` should have worked and did locally - but on the CI providers I still get output without colors codes. Something is going on here that I don't understand yet :/

**Comment by janpio** [comment]
I checked that `Thor::Base.shell.new` does indeed select and "return" `Bundler::Thor::Shell::Color` - and it does. No idea why the output is still missing the color codes then. Makes no sense to me.

**Comment by janpio** [comment]
When running the test directly (`ruby bin/rspec ./spec/bundler/source_spec.rb:66`) or via the file only (`ruby bin/rspec ./spec/bundler/source_spec.rb`) it passes. It fails when running the whole test suite `ruby bin/rspec`.

See the macOS and Linux builds here: https://janpio.visualstudio.com/bundler/_build/results?buildId=548

What could cause this?

**Comment by deivid-rodriguez** [comment]
Hi @janpio! A little late to reply, but the problem you're mentioning should be fixed by #7002. Can you try again your tentative fix (`allow($stdout).to receive(:tty?).and_return(true)`) with that patch applied and see if it works?

**Comment by janpio** [comment]
Sorry didn't get to test this, but seems it have fixed failing tests:

https://github.com/bundler/bundler/runs/72662867
>  412 failed (16.27 %, 0 new, 412 recurring) 
https://github.com/bundler/bundler/runs/73257722
>  409 failed (16.15 %, 0 new, 409 recurring) 

👍
